Cython/Compiler/Parsing.py

```diff
@@ -3968,6 +3975,459 @@ def p_cpp_class_attribute(s, ctx):
+        target = None
+        if s.systring != "_":  # for match-case '_' is treated as a special wildcard
+            target = p_pattern_capture_target(s)
+        else:
+            s.next()
+        pattern = MatchCaseNodes.MatchAndAssignPatternNode(
+            s.position(), target = target, is_star = True
+        )
+        return pattern
+    else:
+        pattern = p_pattern(s)
+        return pattern
+
+
+def p_pattern(s):
+    # try "as_pattern" then "or_pattern"
+    # (but practically "as_pattern" starts with "or_pattern" too)
+    patterns = []
+    pos = s.position()
+    while True:
+        patterns.append(p_closed_pattern(s))
+        if s.sy == "|":
+            s.next()
+        else:
+            break
+
+    if len(patterns) > 1:
+        pattern = MatchCaseNodes.OrPatternNode(
+            pos,
+            alternatives = patterns
+        )
+    else:
+        pattern = patterns[0]
+
+    if s.sy == 'IDENT' and s.systring == 'as':
+        s.next()
+        with tentatively_scan(s) as errors:
+            pattern.as_targets.append(p_pattern_capture_target(s))
+        if errors and s.sy == "_":
+            s.next()
+            # make this a specific error
+            return Nodes.ErrorNode(errors[0].args[0], what = errors[0].args[1])
+        elif errors:
+            with tentatively_scan(s):
+                expr = p_test(s)
+                return Nodes.ErrorNode(expr.pos, what = "Invalid pattern target")
+            s.error(errors[0])
+    return pattern
+
+
+def p_closed_pattern(s):
+    """
+    The PEG parser specifies it as
+    | literal_pattern
+    | capture_pattern
+    | wildcard_pattern
+    | value_pattern
+    | group_pattern
+    | sequence_pattern
+    | mapping_pattern
+    | class_pattern
+
+    For the sake avoiding too much backtracking, we know:
+    * starts with "{" is a sequence_pattern
+    * starts with "[" is a mapping_pattern
+    * starts with "(" is a group_pattern or sequence_pattern
+    * wildcard pattern is just identifier=='_'
+    The rest are then tried in order with backtracking
+    """
+    if s.sy == 'IDENT' and s.systring == '_':
+        pos = s.position()
+        s.next()
+        return MatchCaseNodes.MatchAndAssignPatternNode(pos)
+    elif s.sy == '{':
+        return p_mapping_pattern(s)
+    elif s.sy == '[':
+        return p_sequence_pattern(s)
+    elif s.sy == '(':
+        with tentatively_scan(s) as errors:
+            result = p_group_pattern(s)
+        if not errors:
+            return result
+        return p_sequence_pattern(s)
+
+    with tentatively_scan(s) as errors:
+        result = p_literal_pattern(s)
+    if not errors:
+        return result
+    with tentatively_scan(s) as errors:
+        result = p_capture_pattern(s)
+    if not errors:
+        return result
+    with tentatively_scan(s) as errors:
+        result = p_value_pattern(s)
+    if not errors:
+        return result
+    return p_class_pattern(s)
+
+
+def p_literal_pattern(s):
```

Comment:
Don't we have much of this already?

Reply:
Yes we do. And I'm not hugely happy about the duplication. The question is: are we prepared to add a `in_pattern_match` flag in `p_atom`.

There's some oddities in what's allowed here. For example `-1` is allowed but `+1` isn't. And f-strings are banned while other strings are allowed. Some of that could be moved to post parse checks though.

I'll have a fresh look and see if it can be improved. At the time I wrote it this seemed like the "least worst option"

Reply:
I've factored out two of the large duplicated chunks of `p_atom`. I think that's probably an improvement, but it does make changes to other bits of the parser so is more invasive. It's in a separate commit so can be reverted.

Reply:
> There's some oddities in what's allowed here. For example `-1` is allowed but `+1` isn't. And f-strings are banned while other strings are allowed. Some of that could be moved to post parse checks though.

These checks sound like they shouldn't be done right in the parser. PostParse sounds much better.

Reply:
Maybe we could even get rid of the `ErrorNode` then?

Reply:
> Maybe we could even get rid of the ErrorNode then?

Yeah I've got rid of that and moved the check to post-parse.

> These checks sound like they shouldn't be done right in the parser. PostParse sounds much better.

Not completely sure. At least for the `+` and `-` they're part of the grammar. For the f-strings it looks like CPython catches them in a post-parse validation though